File: ch1/ex07.rb
module Ex07

  def sqrt(x, guess=1.0, prev_guess=0.0)
    if(good_enough?(guess, prev_guess, x))
      guess
    else
      sqrt(x, improve(guess, x), guess)
    end
  end

  private
  def square(x)
    x * x
  end

  def average(x, y)
    (x + y) / 2.0
  end

  def improve(guess, x)
    average(guess, (x / guess))
  end

  def good_enough?(guess, prev_guess, x)
    ((square(guess) - square(prev_guess))/guess).abs < 0.001
  end  

end

extend Ex07

puts "sqrt(4) => #{sqrt(4)}"
puts "sqrt(0.0001) => #{sqrt(0.0001)}"
puts "sqrt(10000000000000000000000000000000000000000000000000000) => #{sqrt(10000000000000000000000000000000000000000000000000000)}"

File: ch1/ex07.js
function square(x) { return x * x }
function average(x, y) { return (x + y) / 2 }
function abs(n) { return n < 0 ? n * -1 : n }
function improve(guess, x) { return average(guess, (x / guess)) }
function isGoodEnough(guess, prevGuess, x) { return abs(((square(guess) - square(prevGuess))/guess)) < 0.001 }
function sqrt(x, guess, prevGuess) {
  if(!guess) { guess = 1.0 }
  if(!prevGuess) { prevGuess = 0.0 }
  if(isGoodEnough(guess, prevGuess, x)) {
    return guess;
  } else {
    return sqrt(x, improve(guess, x), guess);
  }
}